include/deal.II/algorithms/discrete_time.h

```diff
@@ -0,0 +1,129 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_discrete_time_h
+#define dealii_discrete_time_h
+
+#include <deal.II/base/exceptions.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Algorithms
+{
+  /**
+   * Provides a means to keep track of the simulation time in a time-dependent
+   * simulation. It manages stepping forward from an initial time to a final
+   * time. It only allows fixed time step size, although the step size may be
+   * automatically adjusted slightly to capture the final time.
+   *
+   * @author Reza Rastak, 2019
+   */
+  class DiscreteTime
+  {
+  public:
+    /**
+     * Constructor
+     */
+    DiscreteTime(const double start_time,
+                 const double final_time,
+                 const double start_step_size);
+
+    /**
+     * Return the end of the time interval.
+     * The final time step end exactly at this point.
+     */
+    constexpr double
+    final_time() const;
+
+    /**
+     * Return the size of the step from current time step to the next.
+     */
+    constexpr double
+    next_step_size() const;
+
+    /**
+     * Return the current time.
+     */
+    constexpr double
+    current_time() const;
+
+    /**
+     * Advance the current time based on the value of the current step.
+     * Afterward, it modifies the value of the time step size if needed.
+     */
+    void
+    advance_time();
+
+    /**
+     * Set the current time equal to start time and set the step size to the
+     * initial step size.
+     */
+    void
+    restart();
+
+  private:
+    /**
+     * The beginning of the time interval.
+     */
+    const double m_start_time;
```

Comment:
We don't generally use `m_` for member variables. Can we just name these variables `start_time` etc? You'll get into conflict with the member variables, so let's call these `get_xxx()` then.

Reply:
Although I slightly dislike the verbosity of adding `get_` to the member functions, your suggestion makes perfect sense because of more readability and consistency. I will fix it.